EIPS/eip-7702.md

```diff
@@ -40,38 +41,67 @@ We introduce a new [EIP-2718](./eip-2718.md) transaction, "set code transaction"
 ```
 rlp([chain_id, nonce, max_priority_fee_per_gas, max_fee_per_gas, gas_limit, destination, value, data, access_list, authorization_list, signature_y_parity, signature_r, signature_s])
 
-authorization_list = [[chain_id, address, [nonce], y_parity, r, s], ...]
+authorization_list = [[chain_id, address, y_parity, r, s], ...]
 ```
 
 The fields `chain_id`, `nonce`, `max_priority_fee_per_gas`, `max_fee_per_gas`, `gas_limit`, `destination`, `value`, `data`, and `access_list` of the outer transaction follow the same semantics as [EIP-1559](./eip-1559.md).
 
-The `authorization_list` is a list of tuples that store the address to code which the signer desires to set in their EOA temporarily. The third element is a list item mimicking an optional value. When the list length is zero, consider the authorization nonce to be null. When the list length is one, consider the single integer value to be the provided nonce authorization. Other lengths and value types in this optional are invalid.
+The `authorization_list` is a list of tuples that store the address to code which the signer desires to set in their EOA temporarily.
 
 The [EIP-2718](./eip-2718.md) `ReceiptPayload` for this transaction is `rlp([status, cumulative_transaction_gas_used, logs_bloom, logs])`.
 
 #### Behavior
 
-At the start of executing the transaction, for each `[chain_id, address, [nonce], y_parity, r, s]` tuple:
+At the start of executing the transaction, for each `[chain_id, address, y_parity, r, s]` tuple:
 
-1. `authority = ecrecover(keccak(MAGIC || rlp([chain_id, [nonce], address])), y_parity, r, s]`
+1. `authority = ecrecover(keccak(MAGIC || rlp([chain_id, address])), y_parity, r, s)`
 2. Verify the chain id is either 0 or the chain's current ID.
 3. Verify that the code of `authority` is empty.
-4. If nonce list item is length one, verify the nonce of `authority` is equal to `nonce`.
+4. Verify that the storage slot `keccak(authority || address)` of account `AUTH_REVOCATION_ADDRESS` is empty. (`authority` and `address` should be padded to 32 bytes each.)
```

Comment:
(Just for clarity, according to the spec "authority" is the signer. For the authorized address the only term that I see other than "address" is "template" so I'll use that.)

This concern also applies to never-revocable signatures as allowed by the current version of the EIP. Part of my motivation for writing this proposal is that I believe current-nonce-revocable signatures do not address the ecosystem's needs and wallets will predominantly use never-revocable signatures. This would leave the majority of users with the worst outcome.

If revocation is template-based, users at least have some recourse: the new wallet could scan the chain for past authorizations published on chain and offer the user to revoke the templates it finds that have been authorized and not revoked.

The old wallet generating a signature and storing it somewhere else (a server) without ever publishing it is a possibility. However, I don't think this is a particularly new attack vector, as without EIP-7702 the wallet could generate a signed transaction with a nonce far in the future to steal user assets and just wait until the right moment to abuse it.

Reply:
> I believe current-nonce-revocable signatures do not address the ecosystem's needs and wallets will predominantly use never-revocable signatures. This would leave the majority of users with the worst outcome.

I share that concern.  I don't think most wallets will use fixed nonce revocation.

> the new wallet could scan the chain for past authorizations published on chain

1. No way to tell if all past authorizations were published on chain. E.g. maybe the old wallet signed a social recovery contract but it was never used.
2. Not easy to scan since the transactions that used the auth could come from any EOA.  Requires scanning all 7702 transactions from all EOAs.

> However, I don't think this is a particularly new attack vector, as without EIP-7702 the wallet could generate a signed transaction with a nonce far in the future to steal user assets and just wait until the right moment to abuse it.

Right - if we're thinking about an attack by a malicious wallet. But it could be an off-chain signature for a legitimate reason, e.g. social recovery contract.  When you switch to a new contract you probably also want to revoke whatever recovery contract the old wallet used.

But I agree with you that users are better off with this revocation mechanism than with fixed nonce.  I just think something equivalent to `maxNonce` would mitigate the issues above.  With `maxNonce` the new wallet can revoke all past auths, whether they were published or not.  I wonder if we could do the same with a system contract like the one you're proposing.  E.g. storing an `authNonce` in the system contract (`keccak(authNonce || address)`) and replacing the current `nonce` in the `authority` with `authNonce`.  The user can then bump `authNonce` in the system contract to revoke all past auths.

Reply:
> But it could be an off-chain signature for a legitimate reason, e.g. social recovery contract.

Yes this is true and a good use case!

> I wonder if we could do the same with a system contract like the one you're proposing. E.g. storing an `authNonce` in the system contract (`keccak(authNonce || address)`) and replacing the current `nonce` in the `authority` with `authNonce`. The user can then bump `authNonce` in the system contract to revoke all past auths.

Ah, this could work too. Are you proposing to combine that with template-based revocation or to just use nonces?

Reply:
> Ah, this could work too. Are you proposing to combine that with template-based revocation or to just use nonces?

We could do both but it adds complexity.  The mechanism I suggested is basically the `maxNonce` system, except that instead of using the account's nonce it uses a storage slot and the bumping part is implemented as EVM code.

In the future (after verkle) we can stop using the system contract and store `authNonce` in the account itself along with its regular `nonce`, as I think @adietrichs suggested.  But then we'll need a way to bump it.

Reply:
If we go with an independent nonce for authorizations, I think "current nonce" is a better option than "max nonce". I don't see a good way to choose the nonce increment for an authorization or for a revocation.

> With `maxNonce` the new wallet can revoke all past auths, whether they were published or not.

I.e., I don't understand how you can be sure your revoked nonce range covers everything you've signed before.

With "current nonce", revoking would be a nuclear option that gets rid of all previously signed authorizations. This is not ideal but it makes some sense.

Reply:
My trust assumptions have changed a bit: We have to trust our wallets. The same trust assumption can be applied with revocations -- Every smart wallet should have option to list down the authorizations it has captured, and provide option to revoke it. Furthermore, if you use only hardware wallets to generate authorizations, hardware wallet can list/revoke the authorizations that went out, and that is considered safer. 
Even if it comes down to wallets having to harvest all 7702 txs for authorizations (coming from a particular EOA account), it seems okay to me -- wallets do index data or use data providers which index data in particular ways for them, and a "get all authorizations from this account". This can then be used in the new wallet to revoke previous authorizations. The caveat is that it doesn't cover authorizations that weren't published on chain (but this is solved with hardware wallet). 

I'm much more skeptical of the template ecosystem (which is regulated with wallet whitelisting, but users will ignore the wallet's warning and sign non-whitelisted templates). With templates, we're entering into the wild west of general smart contracts, of which the smart wallets are a much-scrutinized, higher-ecosystem-trust subset. 
For example, If users unwittingly authorize something like an upgradeable proxy, that becomes lucrative opportunity for the template provider (by swapping underlying implementation and using authorization to execute arbitrary calls in authority context).

There's desire for having previous wallet authorizations continue to work in a new wallet (as indicated in the 7702 proxy recommendation). Also, Blanket revocation (off previous wallet) is not possible only if i) hardware wallet was not used, and ii) previous wallet maintained offchain authorizations, and iii) previous wallet has gone rogue.

Reply:
> If we go with an independent nonce for authorizations, I think "current nonce" is a better option than "max nonce". Otherwise I don't see a good way to choose the nonce increment for an authorization or for a revocation.
> 
> > With `maxNonce` the new wallet can revoke all past auths, whether they were published or not.
> 
> I.e., I don't understand how you can be sure your revoked nonce range covers everything you've signed before.
> 
> With "current nonce", revoking would be a nuclear option that gets rid of all previously signed authorizations. This is not ideal but it makes some sense.

Just to answer this, the proposal was to divide nonce in two parts: the lower bits represent standard EOA tx counter. While the higher bits are "current operational authentication range". Say the split is 32-32 bits. Then nonce bump (revocation, possibly done by a flag in 7702 tx itself) adds 2^32 to the current nonce. 7702 pre-execution checks if the `higher bits of account's current nonce` == `higher bits of authentication's nonce`. This doesn't need anyone to choose nonce increment.